primitives/session/src/offchain_hashdb.rs

```diff
@@ -0,0 +1,312 @@
+                }
+            })
+            .collect::<BTreeSet<_>>()
+
+        } else {
+
+            let index = self.get_index(session_index);
+            index.into_iter().for_each(move |(derived_key, _prefix)| {
+                offchain::local_storage_set(StorageKind::PERSISTENT, derived_key.as_slice(), &[]);
+            });
+            BTreeSet::new()
+
+        };
+        self.set_index(session_index, index);
+    }
+
+    /// Set the current session index.
+    ///
+    /// Utilized by the implementation of `HashDB` for `Self`.
+    pub fn set_session<E: Encode>(&mut self, session_index: E) {
+        self.session_index = session_index.encode();
+    }
+
+    /// Make `key` known to the persistent tracking index.
+    fn add_to_index(&self, derived_key: &[u8], tree_prefix: &[u8], session_index: &[u8]) {
+        let per_session_key: Vec<u8> = Self::derive_tracking_index_key(session_index);
+        let mut mapping: BTreeSet<(Vec<u8>, Vec<u8>)> =
+            offchain::local_storage_get(StorageKind::PERSISTENT, derived_key)
+                .map(|bytes| {
+                    <BTreeSet<(Vec<u8>, Vec<u8>)> as Decode>::decode(&mut &bytes[..]).unwrap()
+                })
+                .unwrap_or_else(|| BTreeSet::new());
+
+        mapping.insert((derived_key.to_vec(), tree_prefix.to_vec()));
+
+        let encoded = mapping.encode();
+        offchain::local_storage_set(StorageKind::PERSISTENT, per_session_key.as_ref(), encoded.as_slice());
+    }
+
+    /// Forget `key` from the persistent index.
+    fn remove_from_index(&self, derived_key: &[u8], tree_prefix: &[u8], session_index: &[u8]) {
+        let index = self.get_index(session_index);
+        let index = index
+            .into_iter()
+            .filter(|(derived_key2, tree_prefix2)| {
+                tree_prefix2.as_slice() != tree_prefix || derived_key2.as_slice() != derived_key
+            })
+            .collect();
+        self.set_index(session_index, index);
+    }
+
+    /// Get the key tracking index.
+    fn get_index(&self, session_index: &[u8]) -> BTreeSet<(Vec<u8>, Vec<u8>)> {
+        let tracking_key: Vec<u8> = Self::derive_tracking_index_key(session_index);
+        let mapping: BTreeSet<(Vec<u8>, Vec<u8>)> =
+            offchain::local_storage_get(StorageKind::PERSISTENT, tracking_key.as_ref())
+                .map(|bytes| {
+                    <BTreeSet<(Vec<u8>, Vec<u8>)> as Decode>::decode(&mut  &bytes[..]).unwrap()
+                })
+                .unwrap_or_else(|| BTreeSet::new());
+        mapping
+    }
+
+    /// Set the (modified) key tracking index.
+    fn set_index(&self, session_index: &[u8], val: BTreeSet<(Vec<u8>, Vec<u8>)>) {
+        let tracking_key: Vec<u8> = Self::derive_tracking_index_key(session_index);
+        let val = val.encode();
+        offchain::local_storage_set(StorageKind::PERSISTENT, tracking_key.as_ref(), val.as_slice());
+    }
+
+
+    fn derive_tracking_index_key(session_index: &[u8]) -> Vec<u8> {
```

Comment:
could this and other functions with a `self` receiver be free functions rather than inherent impls? Free function is typically more idiomatic in Rust; inherent impls are usually only used as constructors.

Reply:
Could be, the only reason was the move of the const prefixes into the impl, but if both live outside, it'll work - will do